src/cachetools/__init__.py

```diff
@@ -497,78 +499,110 @@ def __getlink(self, key):
         return value
 
 
+def _get_function_value_from_cache(cache, key, lock=None):
+    lock = lock or contextlib.nullcontext()
+    try:
+        with lock:
+            return cache[key]
+    except KeyError:
+        return None
+
+
+def _set_function_value_in_cache(cache, key, value, lock=None):
+    lock = lock or contextlib.nullcontext()
+    try:
+        with lock:
+            cache.setdefault(key, value)
+    except ValueError:
+        pass
+
+
 def cached(cache, key=hashkey, lock=None):
     """Decorator to wrap a function with a memoizing callable that saves
     results in a cache.
 
     """
 
     def decorator(func):
-        if cache is None:
+        if inspect.iscoroutinefunction(func):
 
-            def wrapper(*args, **kwargs):
-                return func(*args, **kwargs)
+            if cache is None:
 
-        elif lock is None:
+                async def wrapper(*args, **kwargs):
+                    return await func(*args, **kwargs)
 
-            def wrapper(*args, **kwargs):
-                k = key(*args, **kwargs)
-                try:
-                    return cache[k]
-                except KeyError:
-                    pass  # key not found
-                v = func(*args, **kwargs)
-                try:
-                    cache[k] = v
-                except ValueError:
-                    pass  # value too large
-                return v
+            else:
+
+                async def wrapper(*args, **kwargs):
+                    k = key(*args, **kwargs)
+                    v = _get_function_value_from_cache(cache, k, lock)
+                    if v:
```

Comment:
If I'm right, this changes the behavior of the decorator to not  cache None/False or values that are interpreted as False.

Reply:
Ah, that's a good point, thanks!

I'll fix the behavior and add a test to ensure it's solid. I'll probably either return a sentinel or indeed revert to using exceptions here, not yet sure which one is the most elegant.

Reply:
I updated my code to use a sentinel, which makes it all nicer. Thanks a lot for the hint!

Reply:
NP. Like your approach & initiative. Did you check the other PR? It will collide with this one.